dev-docs/RFCs/v7.3/program-manager-rfc.md

```diff
@@ -0,0 +1,125 @@
+- Shader hooks and module injections are global state meaning that an application that imports luma.gl must use the same definitions globally (a case where one might not want this is when rendering to separate canvases on the same page).
+- Caching being limited per-model presents serious performance concerns as program compilation is expensive, and program switching is among the more expensive GL state changes (this would be a concern, for example, in deck.gl when rendering multiple instances of the same layer).
+- With out a centralized system for managing programs, it will be difficult to take advantage of optimizations that require batching program operations, such `KHR_parallel_shader_compile`.
+
+
+## Overview
+
+The proposed program manager would take the caching mechanisms provide by `ShaderCache` and extend it in the following ways:
+- Remove the per-model scoping.
+- Manage shader sources, modules, injections and shader hooks.
+- Provide a simple API for building a program from shader sources, modules and defines, returning a cached version if available.
+
+The `BaseModel` class would be modified to:
+- Support attachment of a program manager from which it would acquire programs.
+- Support sharing programs with other `BaseModel`s.
+- Support switching programs.
+
+## Implementation
+
+A `ProgramManager` that supports the following methods:
+- `addVertexShader(id, source)`: register vertex shader source code.
+- `addFragmentShader(id, source)`: register fragment shader source code.
+- `addModule(id, module, injections)`: register a module, along with optional shader hook injections.
+- `addShaderHook(type, {signature, header, footer})`: register a shader hook function.
+- `get(vsId, fsId, {defines, modules})`: return a program. Compile and cache it on first call, return cached version subsequently (and increment usage count).
+- `release(program)`: indicate that program is no longer being used (decrement usage count, delete if it reaches 0).
+
+`BaseModel` would be modified to:
+- Take a `ProgramManager` instance as construction parameter, from which it can get programs.
+- Update its `Program`s uniforms on every draw. This will allow program sharing and program switching.
+- Add a method `updateProgram(vsId, fdId, {defines, modules})` that would allow it to switch programs (via its `ProgramManger`) without having to be rebuilt. 
```

Comment:
Style nit: 

I try to avoid mixing positional and named parameters `addShaderHook(type, {signature, header, footer})` and `updateProgram(vsId, fdId, {defines, modules})` unless the named parameters are really an optional "options" object. 

It is just my experience that when I come back and use my own APIs, my previous choices in this regard tend to seem arbitrary when I mix styles.